addons/account/models/account_account.py

```diff
@@ -886,23 +887,32 @@ def name_create(self, name):
     @api.model_create_multi
     def create(self, vals_list):
         records_list = []
-        for company_ids, vals_list_for_company in itertools.groupby(vals_list, lambda v: v.get('company_ids')):
-            match company_ids:
-                case None:
-                    company = self.env.company
-                case [(Command.LINK, company_id, *_)] | [(Command.SET, 0, [company_id])] | [company_id] if isinstance(company_id, int):
-                    company = self.env['res.company'].browse(company_id)
-                case _:
-                    raise ValueError(_("You may only give an account a single company at creation."))
+        for company_ids, vals_list_for_company in itertools.groupby(vals_list, lambda v: v.get('company_ids', [])):
+            # Determine companies to set on the account. The first one will be used to create the account, the others later added to it.
+            company_ids = self._fields['company_ids'].convert_to_cache(company_ids, self)
+            companies = self.env['res.company'].browse(company_ids) or self.env.company
             cache = set()
             vals_list_for_company = list(vals_list_for_company)
             for vals in vals_list_for_company:
                 if 'prefix' in vals:
                     prefix, digits = vals.pop('prefix'), vals.pop('code_digits')
                     start_code = prefix.ljust(digits - 1, '0') + '1' if len(prefix) < digits else prefix
-                    vals['code'] = self.with_company(company)._search_new_account_code(start_code, cache)
+                    vals['code'] = self.with_company(companies[0])._search_new_account_code(start_code, cache)
                     cache.add(vals['code'])
-            records_list.append(super(AccountAccount, self.with_company(company)).create(vals_list_for_company))
+                elif 'code' not in vals and 'code_by_company' in vals:
+                    vals['code'] = vals['code_by_company'][companies[0].id]
+                vals['company_ids'] = [Command.set(companies[0].ids)]
+            code_by_company_list = [vals.pop('code_by_company', {}) for vals in vals_list_for_company]
+            # Create the accounts with a single company and a single code.
+            new_accounts = super(AccountAccount, self.with_company(companies[0])).create(vals_list_for_company)
```

Comment:
I get a traceback at this line of code when doing the following:
- have a shared account on multiple companies
- having only the first company selected in the company selector
- copy the account
=> Error

Reply:
Thanks :smile: it should be fixed now.
- The error was caused because we were setting fields with `check_company = True` when creating the account. Since at that point the account only has one company, this was breaking the check_company constraint.
- The solution is to defer the write of those fields to when we write all the `company_ids` on the account.
- It was an undeterministic error, due to the fact that we arbitrarily choose the first of the company_ids as the company we use to create the account at first.